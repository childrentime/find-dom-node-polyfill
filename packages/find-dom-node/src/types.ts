import type { ReactInstance } from "react";

export type Input = ReactInstance | null | undefined;
export type Output = Element | null | Text;
export type Fiber = {
  tag: WorkTag;
  key: null | string;
  elementType: any;
  type: any;
  stateNode: any;
  return: Fiber | null;
  child: Fiber | null;
  sibling: Fiber | null;
  index: number;
  ref: null | (((handle: mixed) => void) & { _stringRef: string | null });
  refCleanup: null | (() => void);
  pendingProps: any;
  memoizedProps: any;
  updateQueue: any; // Update with specific type when known
  memoizedState: any;
  dependencies: Dependencies | null;
  mode: TypeOfMode;
  flags: Flags;
  subtreeFlags: Flags;
  deletions: Fiber[] | null;
  lanes: Lanes;
  childLanes: Lanes;
  alternate: Fiber | null;
  actualDuration?: number;
  actualStartTime?: number;
  selfBaseDuration?: number;
  treeBaseDuration?: number;
  _debugInfo?: ReactDebugInfo | null;
  _debugOwner?: ReactComponentInfo | Fiber | null;
  _debugStack?: string | Error | null;
  _debugTask?: ConsoleTask | null;
  _debugIsCurrentlyTiming?: boolean;
  _debugNeedsRemount?: boolean;
  _debugHookTypes?: HookType[] | null;
};

type ReactDebugInfo = any;
type ReactComponentInfo = any;
type ConsoleTask = any;
type mixed = any;
type Dependencies = any;
type TypeOfMode = number;
type Flags = number;
type Lanes = number;
type HookType =
  | "useState"
  | "useReducer"
  | "useContext"
  | "useRef"
  | "useEffect"
  | "useEffectEvent"
  | "useInsertionEffect"
  | "useLayoutEffect"
  | "useCallback"
  | "useMemo"
  | "useImperativeHandle"
  | "useDebugValue"
  | "useDeferredValue"
  | "useTransition"
  | "useSyncExternalStore"
  | "useId"
  | "useCacheRefresh"
  | "useOptimistic"
  | "useFormState"
  | "useActionState";
type WorkTag =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23
  | 24
  | 25
  | 26
  | 27
  | 28
  | 29;
