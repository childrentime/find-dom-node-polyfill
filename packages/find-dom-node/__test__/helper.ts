export function captureConsoleErrors() {
  const originalError = console.error;
  const errors: string[] = [];

  console.error = function (...args: any) {
    originalError.apply(console, args); // Preserve the original behavior
    errors.push(args);
  };

  return {
    getErrors: () => errors,
    restore: () => {
      console.error = originalError;
    }
  };
}