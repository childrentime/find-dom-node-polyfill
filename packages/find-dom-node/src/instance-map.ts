import { Fiber } from "./types";

export function get(key): Fiber {
  return key?._reactInternals;
}