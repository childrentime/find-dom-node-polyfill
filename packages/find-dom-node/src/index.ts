import { Input, Output } from "./types";
import { get as getInstance } from "./instance-map";
import { findCurrentHostFiber } from "./fiber";

export function findDOMNode(component: Input): Output {
  if (component === null) return null;
  const fiber = getInstance(component);
  if (fiber === undefined) {
    // @ts-ignore
    if (typeof component?.render === 'function') {
      throw new Error('Unable to find node on an unmounted component.');
    } else {
      // @ts-ignore
      const keys = Object.keys(component).join(',');
      throw new Error(
        `Argument appears to not be a ReactComponent. Keys: ${keys}`,
      );
    }
  }
  const hostFiber = findCurrentHostFiber(fiber);
  if (hostFiber === null) {
    return null;
  }
  return hostFiber.stateNode;
}