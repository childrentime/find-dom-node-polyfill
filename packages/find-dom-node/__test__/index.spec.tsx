/**
 * @vitest-environment jsdom
 */

import { expect, describe, it } from 'vitest'
import { findDOMNode } from '../src';
import React, { StrictMode } from 'react';
import ReactDOM from 'react-dom';
import { captureConsoleErrors } from './helper';

/**
 * @description copy from findDOMNodeFB.test.js
 */
describe('findDOMNode', () => {
  it('findDOMNode should return null if passed null', () => {
    expect(findDOMNode(null)).toBe(null);
  });

  it('findDOMNode should find dom element', () => {
    class MyNode extends React.Component {
      render() {
        return (
          <div>
            <span>Noise</span>
          </div>
        );
      }
    }

    const container = document.createElement('div');
    const myNode = ReactDOM.render(<MyNode />, container);
    const myDiv = ReactDOM.findDOMNode(myNode);
    const mySameDiv = ReactDOM.findDOMNode(myDiv);
    expect(myDiv.tagName).toBe('DIV');
    expect(mySameDiv).toBe(myDiv);
  });

  it('findDOMNode should find dom element after an update from null', () => {
    function Bar({flag}) {
      if (flag) {
        return <span>A</span>;
      }
      return null;
    }
    class MyNode extends React.Component<{flag?: boolean}> {
      render() {
        return <Bar flag={this.props.flag} />;
      }
    }

    const container = document.createElement('div');

    const myNodeA = ReactDOM.render(<MyNode />, container);
    const a = ReactDOM.findDOMNode(myNodeA);
    expect(a).toBe(null);

    const myNodeB = ReactDOM.render(<MyNode flag={true} />, container);
    expect(myNodeA === myNodeB).toBe(true);

    const b = ReactDOM.findDOMNode(myNodeB);
    expect(b.tagName).toBe('SPAN');
  });

  it('findDOMNode should reject random objects', () => {
    expect(function () {
      ReactDOM.findDOMNode({foo: 'bar'});
    }).toThrowError('Argument appears to not be a ReactComponent. Keys: foo');
  });

  it('findDOMNode should reject unmounted objects with render func', () => {
    class Foo extends React.Component {
      render() {
        return <div />;
      }
    }

    const container = document.createElement('div');
    const inst = ReactDOM.render(<Foo />, container);
    ReactDOM.unmountComponentAtNode(container);

    expect(() => ReactDOM.findDOMNode(inst)).toThrowError(
      'Unable to find node on an unmounted component.',
    );
  });

  it('findDOMNode should not throw an error when called within a component that is not mounted', () => {
    class Bar extends React.Component {
      UNSAFE_componentWillMount() {
        expect(ReactDOM.findDOMNode(this)).toBeNull();
      }

      render() {
        return <div />;
      }
    }
    expect(() => {
      const container = document.createElement('div');
      ReactDOM.render(<Bar />, container);
    }).not.toThrow();
  });

  it('findDOMNode should warn if used to find a host component inside StrictMode', () => {
    let parent = undefined;
    let child = undefined;

    class ContainsStrictModeChild extends React.Component {
      render() {
        return (
          <StrictMode>
            <div ref={n => (child = n)} />
          </StrictMode>
        );
      }
    }

    const container = document.createElement('div');
    ReactDOM.render(
      <ContainsStrictModeChild ref={n => (parent = n)} />,
      container,
    );

    const errors = captureConsoleErrors();
    let match = ReactDOM.findDOMNode(parent);
    expect(errors.getErrors()).toMatchSnapshot();
    expect(match).toBe(child);
  });

  it('findDOMNode should warn if passed a component that is inside StrictMode', () => {
    let parent = undefined;
    let child = undefined;

    class IsInStrictMode extends React.Component {
      render() {
        return <div ref={n => (child = n)} />;
      }
    }

    const container = document.createElement('div');

    ReactDOM.render(
      <StrictMode>
        <IsInStrictMode ref={n => (parent = n)} />
      </StrictMode>,
      container,
    );

    const errors = captureConsoleErrors();
    let match = ReactDOM.findDOMNode(parent);
    expect(errors.getErrors()).toMatchSnapshot();
    expect(match).toBe(child);
  });
})