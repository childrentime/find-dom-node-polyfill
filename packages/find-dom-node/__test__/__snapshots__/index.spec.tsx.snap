// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findDOMNode > findDOMNode should warn if passed a component that is inside StrictMode 1`] = `
[
  [
    "Warning: %s is deprecated in StrictMode. %s was passed an instance of %s which is inside StrictMode. Instead, add a ref directly to the element you want to reference. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-find-node%s",
    "findDOMNode",
    "findDOMNode",
    "IsInStrictMode",
    "
    at div
    at IsInStrictMode (/Users/zhangyuanqing/works/github/find-dom-node-polyfill/packages/find-dom-node/__test__/index.spec.tsx:102:5)",
  ],
]
`;

exports[`findDOMNode > findDOMNode should warn if used to find a host component inside StrictMode 1`] = `
[
  [
    "Warning: %s is deprecated in StrictMode. %s was passed an instance of %s which renders StrictMode children. Instead, add a ref directly to the element you want to reference. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-find-node%s",
    "findDOMNode",
    "findDOMNode",
    "ContainsStrictModeChild",
    "
    at div
    at ContainsStrictModeChild (/Users/zhangyuanqing/works/github/find-dom-node-polyfill/packages/find-dom-node/__test__/index.spec.tsx:84:5)",
  ],
]
`;
